programs/cp-amm/src/manager/swap_manager.rs

```diff
@@ -0,0 +1,91 @@
+use crate::{
+    activation_handler::ActivationHandler,
+    constants::fee::{FEE_DENOMINATOR, MAX_BASIS_POINT, MAX_FEE_NUMERATOR},
+    get_pool_access_validator_without_clock,
+    params::swap::TradeDirection,
+    safe_math::SafeMath,
+    state::{fee::FeeMode, Pool, SwapResult},
+    PoolError,
+};
+use anchor_lang::prelude::*;
+
+pub struct SwapFinalResult {
+    pub out_amount: u64,
+    pub fee_amount: u64,
+    pub fee_mint: Pubkey,
+    pub fee_bps: u128,
+}
+
+pub fn swap(
+    pool: Pool,
+    amount_in: u64,
+    current_slot: u64,
+    current_timestamp: i64,
+    a_to_b: bool,
+) -> Result<SwapFinalResult> {
+    let access_validator =
+        get_pool_access_validator_without_clock(&pool, current_slot, current_timestamp)?;
+
+    require!(
+        access_validator.can_swap(&Pubkey::default()),
+        PoolError::PoolDisabled
+    );
+
+    handle_normal_swap(pool, amount_in, current_slot, current_timestamp, a_to_b)
+}
+
+fn handle_normal_swap(
+    mut pool: Pool,
+    amount_in: u64,
+    current_slot: u64,
+    current_timestamp: i64,
+    a_to_b: bool,
+) -> Result<SwapFinalResult> {
+    // update for dynamic fee reference
+    pool.update_pre_swap(current_timestamp as u64)?;
+
+    let trade_direction = if a_to_b {
+        TradeDirection::AtoB
+    } else {
+        TradeDirection::BtoA
+    };
+
+    let current_point = ActivationHandler::get_current_point_without_clock(
+        pool.activation_type,
+        current_slot,
+        current_timestamp,
+    )?;
+
+    let fee_mode = FeeMode::get_fee_mode(pool.collect_fee_mode, trade_direction, false)?;
```

Comment:
no referral?

Reply:
Jupiter quote doesn't require referral fee amount, it only require the trade fee amount charged for the swap.